### PR TITLE
Add spec for `BLOCKHASH` opcode

### PR DESCRIPTION
**Pull request by jeong0982**
Not sure for handling valid range. If there is any appropriate way for checking valid range, please let me know.
Thank you.

### CODE REVIEW

File: src/zkevm_specs/evm/execution/blockhash.py
Comment (roynalnaruto) [inline src/zkevm_specs/evm/execution/blockhash.py:19]: You probably can't use the `Range256` variant in `FixedTable` to check whether or not the difference between `current_block_number` and `block_number` is `< 256`. Mainly because you want to handle the case where the lookup fails, i.e. simply push `0` to the stack.

So I think you will need a comparison gadget to check that `block_number < current_block_number`.

So you would have something like:
```python
# get value that was pushed to stack (RLC-encoded)
block_hash = instruction.stack_push()

# comparing block_number and current_block_number
block_lt, _ = instruction.compare(block_number, current_block_number, 8)
diff_lt, _ = instruction.compare(current_block_number - block_number, FQ(256), 2)

# get the expected block hash depending on the above conditions (RLC-encoded)
expected_block_hash = instruction.select(
    block_lt * diff_lt,
    instruction.block_context_lookup(
        BlockContextFieldTag.HistoryHash,
        block_number,
    ),
    FQ.zero(),
)

# block hash is as expected
instruction.constrain_equal(block_hash, expected_block_hash)
```

**Review by roynalnaruto — CHANGES_REQUESTED** [review]
Mostly looks good. I have posted some comments/suggestions :)

File: src/zkevm_specs/evm/execution/blockhash.py
Comment (jeong0982) [reply src/zkevm_specs/evm/execution/blockhash.py:19]: That's exactly what I want to know. Thank you for your comment!

Comment (jeong0982) [reply src/zkevm_specs/evm/execution/blockhash.py:19]: @roynalnaruto There is an error from `block_conext_lookup()` in `select` instruction. Because `instruction.select()` function is called after calculating argument, so `block_context_lookup` with invalid `block_number` raise `LookupUnsatFailure`. Do you have any idea for handling this situation? Thank you.

Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/blockhash.py:19]: Ah, of course. Yes that's expected.

In that case, you could scrap the `instruction.select(..)` and simply have:
```python
if block_lt * diff_lt == FQ.one():
    expected_block_hash = instruction.block_context_lookup(
        BlockContextFieldTag.HistoryHash,
        block_number,
    ),
else:
    expected_block_hash = FQ.zero()
```

Comment (jeong0982) [reply src/zkevm_specs/evm/execution/blockhash.py:19]: Thanks a lot 👍🏼

Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/blockhash.py]: @jeong0982 you will need to fix the formatting:

* To check which file needs formatting
```
$ make lint
```
* Formatting
```
$ black src/
$ black tests/
```

Comment (jeong0982) [reply src/zkevm_specs/evm/execution/blockhash.py]: Ok. Thank you!

File: specs/opcode/40BLOCKHASH.md
Comment (roynalnaruto) [inline specs/opcode/40BLOCKHASH.md:20]: This needs to be removed now

**Review by roynalnaruto — APPROVED** [review]
LGTM! But we will wait for other reviewers to review as well!

File: specs/opcode/40BLOCKHASH.md
Comment (jeong0982) [reply specs/opcode/40BLOCKHASH.md]: Fixed. Thanks!

**Review by CPerezz — APPROVED** [review]
LGTM modulo the comment I added.

Thanks for the work!

File: specs/opcode/40BLOCKHASH.md
Comment (roynalnaruto) [reply specs/opcode/40BLOCKHASH.md]: Oops! yes of course it _should_ be included. When I saw the markdown file on a small screen, due to some indentation  issues I thought the `block_number` was included within `lookups` (hence I mentioned that this is not a lookup).

@jeong0982 since this is a constraint, yes it should be present in the spec doc. Thanks for noticing @CPerezz :)

Comment (jeong0982) [reply specs/opcode/40BLOCKHASH.md]: I was also confused. I will put it back. Thank you👍🏼

File: specs/opcode/40BLOCKHASH.md
Comment (jeong0982) [reply specs/opcode/40BLOCKHASH.md]: Fixed. Thanks!

**Comment by roynalnaruto** [comment]
Thank you @jeong0982 for the contribution! I will merge this since it has 2 approvals and all comments/discussions have been resolved!

**Pull request merged**
